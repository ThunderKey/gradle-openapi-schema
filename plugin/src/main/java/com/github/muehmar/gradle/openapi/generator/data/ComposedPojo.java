package com.github.muehmar.gradle.openapi.generator.data;

import ch.bluecare.commons.data.PList;

import java.util.Objects;

/**
 * Represents a composition of other definitions, i.e. pojos. Distinguishes between different types
 * of compositions, see {@link CompositionType}.
 */
public class ComposedPojo {
  private final CompositionType type;
  private final PList<String> pojoNames;
  private final PList<PojoMember> members;

  public enum CompositionType {
    ALL_OF,
    ANY_OF,
    ONE_OF;
  }

  public ComposedPojo(CompositionType type, PList<String> pojoNames, PList<PojoMember> members) {
    this.type = type;
    this.pojoNames = pojoNames;
    this.members = members;
  }

  public CompositionType getType() {
    return type;
  }

  public PList<String> getPojoNames() {
    return pojoNames;
  }

  public PList<PojoMember> getMembers() {
    return members;
  }

  @Override
  public boolean equals(Object o) {

    if (this == o) {
      return true;
    }
    if (o == null || getClass() != o.getClass()) {
      return false;
    }
    ComposedPojo that = (ComposedPojo) o;
    return type == that.type
        && Objects.equals(pojoNames, that.pojoNames)
        && Objects.equals(members, that.members);
  }

  @Override
  public int hashCode() {
    return Objects.hash(type, pojoNames, members);
  }

  @Override
  public String toString() {
    return "ComposedPojo{"
        + "type="
        + type
        + ", pojoNames="
        + pojoNames
        + ", members="
        + members
        + '}';
  }
}
