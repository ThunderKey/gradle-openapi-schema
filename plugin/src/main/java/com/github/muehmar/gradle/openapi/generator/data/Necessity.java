package com.github.muehmar.gradle.openapi.generator.data;

/** Defines the necessity of an attribute, i.e. if the attribute is required or optional. */
public enum Necessity {
  REQUIRED,
  OPTIONAL;

  public static Necessity fromBoolean(boolean necessity) {
    return necessity ? REQUIRED : OPTIONAL;
  }
}
