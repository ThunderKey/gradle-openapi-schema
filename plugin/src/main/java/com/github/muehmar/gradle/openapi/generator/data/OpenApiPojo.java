package com.github.muehmar.gradle.openapi.generator.data;

import io.swagger.v3.oas.models.media.Schema;
import java.util.Objects;

/** Container holding a key and the corresponding {@link Schema}. */
public class OpenApiPojo {
  private final Name name;
  private final Schema<?> schema;

  public OpenApiPojo(Name name, Schema<?> schema) {
    this.name = name;
    this.schema = schema;
  }

  public Name getName() {
    return name;
  }

  @SuppressWarnings("java:S1452")
  public Schema<?> getSchema() {
    return schema;
  }

  @Override
  public boolean equals(Object o) {
    if (this == o) {
      return true;
    }
    if (o == null || getClass() != o.getClass()) {
      return false;
    }
    OpenApiPojo that = (OpenApiPojo) o;
    return Objects.equals(name, that.name) && Objects.equals(schema, that.schema);
  }

  @Override
  public int hashCode() {
    return Objects.hash(name, schema);
  }

  @Override
  public String toString() {
    return "OpenApiPojo{" + "name='" + name + '\'' + ", schema=" + schema + '}';
  }
}
