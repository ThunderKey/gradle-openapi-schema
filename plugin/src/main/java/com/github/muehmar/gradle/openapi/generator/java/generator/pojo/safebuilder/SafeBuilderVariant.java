package com.github.muehmar.gradle.openapi.generator.java.generator.pojo.safebuilder;

public enum SafeBuilderVariant {
  STANDARD(""),
  FULL("Full");

  private final String builderNamePrefix;

  SafeBuilderVariant(String builderNamePrefix) {
    this.builderNamePrefix = builderNamePrefix;
  }

  public String getBuilderNamePrefix() {
    return builderNamePrefix;
  }
}
