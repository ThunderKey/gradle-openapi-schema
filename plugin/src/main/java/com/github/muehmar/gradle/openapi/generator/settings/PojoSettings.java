package com.github.muehmar.gradle.openapi.generator.settings;

import ch.bluecare.commons.data.PList;
import io.github.muehmar.pojoextension.annotations.PojoExtension;
import java.io.Serializable;
import java.util.List;
import java.util.Objects;

@PojoExtension
public class PojoSettings extends PojoSettingsExtension implements Serializable {
  private final JsonSupport jsonSupport;
  private final String packageName;
  private final String suffix;
  private final boolean enableSafeBuilder;
  private final boolean enableConstraints;
  private final List<ClassTypeMapping> classTypeMappings;
  private final List<FormatTypeMapping> formatTypeMappings;
  private final EnumDescriptionSettings enumDescriptionSettings;

  PojoSettings(
      JsonSupport jsonSupport,
      String packageName,
      String suffix,
      boolean enableSafeBuilder,
      boolean enableConstraints,
      List<ClassTypeMapping> classTypeMappings,
      List<FormatTypeMapping> formatTypeMappings,
      EnumDescriptionSettings enumDescriptionSettings) {
    this.jsonSupport = jsonSupport;
    this.packageName = packageName;
    this.suffix = suffix;
    this.enableSafeBuilder = enableSafeBuilder;
    this.enableConstraints = enableConstraints;
    this.classTypeMappings = classTypeMappings;
    this.formatTypeMappings = formatTypeMappings;
    this.enumDescriptionSettings = enumDescriptionSettings;
  }

  public PojoSettings(
      JsonSupport jsonSupport,
      String packageName,
      String suffix,
      boolean enableSafeBuilder,
      boolean enableConstraints,
      PList<ClassTypeMapping> classTypeMappings,
      PList<FormatTypeMapping> formatTypeMappings,
      EnumDescriptionSettings enumDescriptionSettings) {
    this.jsonSupport = jsonSupport;
    this.packageName = packageName;
    this.suffix = suffix;
    this.enableSafeBuilder = enableSafeBuilder;
    this.enableConstraints = enableConstraints;
    this.classTypeMappings = classTypeMappings.toArrayList();
    this.formatTypeMappings = formatTypeMappings.toArrayList();
    this.enumDescriptionSettings = enumDescriptionSettings;
  }

  public String getPackageName() {
    return packageName;
  }

  public String getSuffix() {
    return suffix;
  }

  public JsonSupport getJsonSupport() {
    return jsonSupport;
  }

  public boolean isJacksonJson() {
    return jsonSupport.equals(JsonSupport.JACKSON);
  }

  public boolean isEnableConstraints() {
    return enableConstraints;
  }

  public PList<ClassTypeMapping> getClassTypeMappings() {
    return PList.fromIter(classTypeMappings);
  }

  public PList<FormatTypeMapping> getFormatTypeMappings() {
    return PList.fromIter(formatTypeMappings);
  }

  public boolean isEnableSafeBuilder() {
    return enableSafeBuilder;
  }

  public boolean isDisableSafeBuilder() {
    return !isEnableSafeBuilder();
  }

  public EnumDescriptionSettings getEnumDescriptionSettings() {
    return enumDescriptionSettings;
  }

  @Override
  public boolean equals(Object o) {
    if (this == o) {
      return true;
    }
    if (o == null || getClass() != o.getClass()) {
      return false;
    }
    PojoSettings that = (PojoSettings) o;
    return enableSafeBuilder == that.enableSafeBuilder
        && enableConstraints == that.enableConstraints
        && jsonSupport == that.jsonSupport
        && Objects.equals(packageName, that.packageName)
        && Objects.equals(suffix, that.suffix)
        && Objects.equals(classTypeMappings, that.classTypeMappings)
        && Objects.equals(formatTypeMappings, that.formatTypeMappings)
        && Objects.equals(enumDescriptionSettings, that.enumDescriptionSettings);
  }

  @Override
  public int hashCode() {
    return Objects.hash(
        jsonSupport,
        packageName,
        suffix,
        enableSafeBuilder,
        enableConstraints,
        classTypeMappings,
        formatTypeMappings,
        enumDescriptionSettings);
  }

  @Override
  public String toString() {
    return "PojoSettings{"
        + "jsonSupport="
        + jsonSupport
        + ", packageName='"
        + packageName
        + '\''
        + ", suffix='"
        + suffix
        + '\''
        + ", enableSafeBuilder="
        + enableSafeBuilder
        + ", enableConstraints="
        + enableConstraints
        + ", classTypeMappings="
        + classTypeMappings
        + ", formatTypeMappings="
        + formatTypeMappings
        + ", enumDescriptionExtraction="
        + enumDescriptionSettings
        + '}';
  }
}
