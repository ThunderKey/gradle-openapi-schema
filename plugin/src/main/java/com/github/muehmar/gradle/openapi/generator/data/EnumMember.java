package com.github.muehmar.gradle.openapi.generator.data;

import ch.bluecare.commons.data.PList;
import ch.bluecare.commons.data.Pair;
import com.github.muehmar.gradle.openapi.generator.settings.EnumDescriptionSettings;
import java.util.Objects;
import java.util.Optional;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

/** An enum member containing the name itself and optionally a description. */
public class EnumMember {
  private final Name name;
  private final String description;

  public EnumMember(Name name, String description) {
    this.name = name;
    this.description = description;
  }

  /** */
  public static Optional<EnumMember> extractDescription(
      Name memberName, EnumDescriptionSettings settings, String description) {
    if (settings.isDisabled()) {
      return Optional.empty();
    }

    final String prefixMatcher = settings.getPrefixMatcherForMember(memberName);
    final Pattern pattern = Pattern.compile(".*" + prefixMatcher + "(.*)");

    return PList.fromArray(description.split("\n"))
        .map(pattern::matcher)
        .filter(Matcher::find)
        .map(matcher -> matcher.group(1))
        .headOption()
        .map(desc -> new EnumMember(memberName, desc.trim()));
  }

  /**
   * Returns a list of enums with found description. If no description is found, the description
   * will be set to an empty string in case failOnMissingDescription is turned off. In case it is
   * turned on, this method will fail if a description of a member is not found.
   */
  public static PList<EnumMember> extractDescriptions(
      PList<Name> enumMembers, EnumDescriptionSettings settings, String description) {
    if (settings.isDisabled()) {
      return enumMembers.map(name -> new EnumMember(name, ""));
    }

    final PList<Optional<EnumMember>> enumWithDescriptions =
        enumMembers.map(m -> extractDescription(m, settings, description));

    int countOfMembersWithDescription = enumWithDescriptions.flatMapOptional(e -> e).size();
    final boolean incompleteDescription =
        countOfMembersWithDescription != enumMembers.size() && countOfMembersWithDescription != 0;

    if (settings.isFailOnIncompleteDescriptions() && incompleteDescription) {
      final String missingEnumMembers =
          enumWithDescriptions
              .zip(enumMembers)
              .filter(p -> !p.first().isPresent())
              .map(Pair::second)
              .map(Name::asString)
              .mkString(", ");

      final String message =
          String.format(
              "No description found for the following enum members [%s]. "
                  + "Either complete the description for these members or turn off the "
                  + "failOnMissingDescription setting for the enum extraction.",
              missingEnumMembers);
      throw new IllegalStateException(message);
    } else {
      return enumWithDescriptions
          .zip(enumMembers)
          .map(p -> p.first().orElse(new EnumMember(p.second(), "")));
    }
  }

  public Name getName() {
    return name;
  }

  public String getDescription() {
    return description;
  }

  @Override
  public boolean equals(Object o) {
    if (this == o) {
      return true;
    }
    if (o == null || getClass() != o.getClass()) {
      return false;
    }
    EnumMember that = (EnumMember) o;
    return Objects.equals(name, that.name) && Objects.equals(description, that.description);
  }

  @Override
  public int hashCode() {
    return Objects.hash(name, description);
  }

  @Override
  public String toString() {
    return "EnumMember{" + "name=" + name + ", description='" + description + '\'' + '}';
  }
}
