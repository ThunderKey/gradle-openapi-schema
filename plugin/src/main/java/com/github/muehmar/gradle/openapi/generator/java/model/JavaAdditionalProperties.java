package com.github.muehmar.gradle.openapi.generator.java.model;

import com.github.muehmar.gradle.openapi.generator.java.generator.enumpojo.EnumContentBuilder;
import com.github.muehmar.gradle.openapi.generator.java.generator.enumpojo.EnumGenerator;
import com.github.muehmar.gradle.openapi.generator.java.model.type.JavaAnyType;
import com.github.muehmar.gradle.openapi.generator.java.model.type.JavaEnumType;
import com.github.muehmar.gradle.openapi.generator.java.model.type.JavaType;
import com.github.muehmar.gradle.openapi.generator.model.AdditionalProperties;
import com.github.muehmar.gradle.openapi.generator.settings.TypeMappings;
import java.util.Optional;
import java.util.function.Function;
import lombok.Value;

@Value
public class JavaAdditionalProperties {
  private static final JavaIdentifier MAP_PROPERTY_NAME =
      JavaIdentifier.fromString("additionalProperties");
  boolean allowed;
  JavaType type;

  public static JavaAdditionalProperties wrap(
      AdditionalProperties additionalProperties, TypeMappings typeMappings) {
    return new JavaAdditionalProperties(
        additionalProperties.isAllowed(),
        JavaType.wrap(additionalProperties.getType(), typeMappings));
  }

  public static JavaAdditionalProperties anyTypeAllowed() {
    return new JavaAdditionalProperties(true, JavaAnyType.create());
  }

  public static JavaAdditionalProperties allowedFor(JavaType type) {
    return new JavaAdditionalProperties(true, type);
  }

  public static JavaAdditionalProperties notAllowed() {
    return new JavaAdditionalProperties(false, JavaAnyType.create());
  }

  public static JavaIdentifier getPropertyName() {
    return MAP_PROPERTY_NAME;
  }

  public Optional<EnumGenerator.EnumContent> asEnumContent() {
    final Function<JavaEnumType, Optional<EnumGenerator.EnumContent>> createEnumContent =
        enumType ->
            Optional.of(
                EnumContentBuilder.create()
                    .className(JavaIdentifier.fromName(type.getClassName()))
                    .description("Additional property enum")
                    .members(enumType.getMembers())
                    .build());
    return type.fold(
        ignore -> Optional.empty(),
        ignore -> Optional.empty(),
        createEnumContent,
        ignore -> Optional.empty(),
        ignore -> Optional.empty(),
        ignore -> Optional.empty(),
        ignore -> Optional.empty(),
        ignore -> Optional.empty(),
        ignore -> Optional.empty());
  }
}
