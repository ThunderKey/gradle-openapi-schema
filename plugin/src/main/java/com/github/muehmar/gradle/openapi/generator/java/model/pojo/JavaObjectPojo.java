package com.github.muehmar.gradle.openapi.generator.java.model.pojo;

import static com.github.muehmar.gradle.openapi.util.Booleans.not;

import ch.bluecare.commons.data.PList;
import com.github.muehmar.gradle.openapi.exception.OpenApiGeneratorException;
import com.github.muehmar.gradle.openapi.generator.java.generator.pojo.MemberContentBuilder;
import com.github.muehmar.gradle.openapi.generator.java.generator.pojo.MemberGenerator;
import com.github.muehmar.gradle.openapi.generator.java.generator.pojo.WitherContentBuilder;
import com.github.muehmar.gradle.openapi.generator.java.generator.pojo.WitherGenerator;
import com.github.muehmar.gradle.openapi.generator.java.generator.shared.pojo.ConstructorContentBuilder;
import com.github.muehmar.gradle.openapi.generator.java.generator.shared.pojo.EqualsContentBuilder;
import com.github.muehmar.gradle.openapi.generator.java.generator.shared.pojo.EqualsGenerator;
import com.github.muehmar.gradle.openapi.generator.java.generator.shared.pojo.HashCodeContentBuilder;
import com.github.muehmar.gradle.openapi.generator.java.generator.shared.pojo.HashCodeGenerator;
import com.github.muehmar.gradle.openapi.generator.java.generator.shared.pojo.PojoConstructorGenerator;
import com.github.muehmar.gradle.openapi.generator.java.generator.shared.pojo.ToStringContentBuilder;
import com.github.muehmar.gradle.openapi.generator.java.generator.shared.pojo.ToStringGenerator;
import com.github.muehmar.gradle.openapi.generator.java.model.JavaAdditionalProperties;
import com.github.muehmar.gradle.openapi.generator.java.model.JavaIdentifier;
import com.github.muehmar.gradle.openapi.generator.java.model.JavaName;
import com.github.muehmar.gradle.openapi.generator.java.model.JavaPojoMember;
import com.github.muehmar.gradle.openapi.generator.java.model.JavaPojoName;
import com.github.muehmar.gradle.openapi.generator.java.model.PojoType;
import com.github.muehmar.gradle.openapi.generator.java.model.composition.JavaAllOfComposition;
import com.github.muehmar.gradle.openapi.generator.java.model.composition.JavaAnyOfComposition;
import com.github.muehmar.gradle.openapi.generator.java.model.composition.JavaOneOfComposition;
import com.github.muehmar.gradle.openapi.generator.model.constraints.Constraints;
import com.github.muehmar.gradle.openapi.generator.model.pojo.ObjectPojo;
import com.github.muehmar.gradle.openapi.generator.settings.TypeMappings;
import io.github.muehmar.pojobuilder.annotations.PojoBuilder;
import java.util.Comparator;
import java.util.Optional;
import java.util.function.Function;
import lombok.EqualsAndHashCode;
import lombok.ToString;

@PojoBuilder(packagePrivateBuilder = true)
@EqualsAndHashCode
@ToString
public class JavaObjectPojo implements JavaPojo {
  private final JavaPojoName name;
  private final String description;
  private final PList<JavaPojoMember> members;
  private final Optional<JavaAllOfComposition> allOfComposition;
  private final Optional<JavaOneOfComposition> oneOfComposition;
  private final Optional<JavaAnyOfComposition> anyOfComposition;
  private final PojoType type;
  private final PList<JavaRequiredAdditionalProperty> requiredAdditionalProperties;
  private final JavaAdditionalProperties additionalProperties;
  private final Constraints constraints;

  JavaObjectPojo(
      JavaPojoName name,
      String description,
      PList<JavaPojoMember> members,
      Optional<JavaAllOfComposition> allOfComposition,
      Optional<JavaOneOfComposition> oneOfComposition,
      Optional<JavaAnyOfComposition> anyOfComposition,
      PojoType type,
      PList<JavaRequiredAdditionalProperty> requiredAdditionalProperties,
      JavaAdditionalProperties additionalProperties,
      Constraints constraints) {
    this.name = name;
    this.description = Optional.ofNullable(description).orElse("");
    this.members = members;
    this.allOfComposition = allOfComposition;
    this.oneOfComposition = oneOfComposition;
    this.anyOfComposition = anyOfComposition;
    this.type = type;
    this.requiredAdditionalProperties = requiredAdditionalProperties;
    this.additionalProperties = additionalProperties;
    this.constraints = constraints;
    assertPropertiesHaveNotSameNameAndDifferentAttributes(
        name, members, allOfComposition, oneOfComposition, anyOfComposition);
  }

  private static void assertPropertiesHaveNotSameNameAndDifferentAttributes(
      JavaPojoName name,
      PList<JavaPojoMember> members,
      Optional<JavaAllOfComposition> allOfComposition,
      Optional<JavaOneOfComposition> oneOfComposition,
      Optional<JavaAnyOfComposition> anyOfComposition) {
    final PList<JavaPojoMember> allMembers =
        members
            .concat(allOfComposition.map(JavaAllOfComposition::getMembers).orElseGet(PList::empty))
            .concat(oneOfComposition.map(JavaOneOfComposition::getMembers).orElseGet(PList::empty))
            .concat(anyOfComposition.map(JavaAnyOfComposition::getMembers).orElseGet(PList::empty));

    final PList<JavaPojoMember> invalidMembers =
        allMembers
            .filter(
                member1 ->
                    allMembers.exists(member2 -> sameNameButDifferentAttributes(member1, member2)))
            .duplicates(Comparator.comparing(m -> m.getName().asString()));

    if (invalidMembers.nonEmpty()) {
      throw new OpenApiGeneratorException(
          String.format(
              "Cannot create DTO %s: Two or more properties (defined either in properties or in one of the schemas used "
                  + "for a composition (allOf, oneOf or anyOf) have the same name but different attributes (e.g. type, "
                  + "constraints)! Invalid properties: [%s].",
              name, invalidMembers.map(JavaPojoMember::getName).mkString(", ")));
    }
  }

  private static boolean sameNameButDifferentAttributes(JavaPojoMember m1, JavaPojoMember m2) {
    return m1.getName().equals(m2.getName())
        && not(m1.asObjectMember().equals(m2.asObjectMember()));
  }

  public static JavaPojoWrapResult wrap(ObjectPojo objectPojo, TypeMappings typeMappings) {
    if (objectPojo.containsNoneDefaultPropertyScope()) {
      return JavaPojoWrapResultBuilder.create()
          .defaultPojo(createForType(objectPojo, typeMappings, PojoType.DEFAULT))
          .andAllOptionals()
          .requestPojo(createForType(objectPojo, typeMappings, PojoType.REQUEST))
          .responsePojo(createForType(objectPojo, typeMappings, PojoType.RESPONSE))
          .build();
    } else {
      return JavaPojoWrapResult.ofDefaultPojo(
          createForType(objectPojo, typeMappings, PojoType.DEFAULT));
    }
  }

  private static JavaObjectPojo createForType(
      ObjectPojo objectPojo, TypeMappings typeMappings, PojoType type) {
    final PList<JavaPojoMember> members =
        objectPojo
            .getMembers()
            .filter(member -> type.includesPropertyScope(member.getPropertyScope()))
            .map(member -> JavaPojoMember.wrap(member, typeMappings));
    final JavaAdditionalProperties javaAdditionalProperties =
        JavaAdditionalProperties.wrap(objectPojo.getAdditionalProperties(), typeMappings);
    return new JavaObjectPojo(
        JavaPojoName.wrap(type.mapName(objectPojo.getName())),
        objectPojo.getDescription(),
        members,
        objectPojo
            .getAllOfComposition()
            .map(comp -> JavaAllOfComposition.wrap(comp, type, typeMappings)),
        objectPojo
            .getOneOfComposition()
            .map(comp -> JavaOneOfComposition.wrap(comp, type, typeMappings)),
        objectPojo
            .getAnyOfComposition()
            .map(comp -> JavaAnyOfComposition.wrap(comp, type, typeMappings)),
        type,
        objectPojo
            .getRequiredAdditionalProperties()
            .map(
                propName ->
                    JavaRequiredAdditionalProperty.fromNameAndType(
                        propName, javaAdditionalProperties.getType())),
        javaAdditionalProperties,
        objectPojo.getConstraints());
  }

  @Override
  public JavaName getSchemaName() {
    return JavaName.fromName(name.getSchemaName());
  }

  @Override
  public JavaIdentifier getClassName() {
    return name.asJavaName().asIdentifier();
  }

  public JavaIdentifier prefixedClassNameForMethod(String prefix) {
    if (prefix.isEmpty()) {
      return name.asJavaName().startLowerCase().asIdentifier();
    } else {
      return name.asJavaName().startUpperCase().prefix(prefix).asIdentifier();
    }
  }

  @Override
  public String getDescription() {
    return description;
  }

  @Override
  public PojoType getType() {
    return type;
  }

  public boolean isSimpleMapPojo() {
    return members.isEmpty()
        && not(allOfComposition.isPresent())
        && not(oneOfComposition.isPresent())
        && not(anyOfComposition.isPresent());
  }

  public boolean hasRequiredMembers() {
    return members.exists(JavaPojoMember::isRequired);
  }

  public boolean hasNotRequiredMembers() {
    return not(hasRequiredMembers());
  }

  public PList<JavaPojoMember> getMembers() {
    return members;
  }

  private PList<JavaPojoMember> getAllOfMembers() {
    return allOfComposition.map(JavaAllOfComposition::getMembers).orElseGet(PList::empty);
  }

  private PList<JavaPojoMember> getOneOfMembers() {
    return oneOfComposition.map(JavaOneOfComposition::getMembers).orElseGet(PList::empty);
  }

  private PList<JavaPojoMember> getAnyOfMembers() {
    return anyOfComposition.map(JavaAnyOfComposition::getMembers).orElseGet(PList::empty);
  }

  public PList<JavaPojoMember> getAllMembersForComposition() {
    return getMembers()
        .map(member -> member.asInnerEnumOf(getClassName()))
        .concat(getComposedMembers());
  }

  public PList<JavaPojoMember> getComposedMembers() {
    return getAllOfMembers()
        .concat(getOneOfMembers())
        .concat(getAnyOfMembers())
        .distinct(Function.identity());
  }

  public PList<JavaPojoMember> getAllMembers() {
    return members.concat(getComposedMembers()).distinct(Function.identity());
  }

  public Optional<JavaAllOfComposition> getAllOfComposition() {
    return allOfComposition;
  }

  public Optional<JavaOneOfComposition> getOneOfComposition() {
    return oneOfComposition;
  }

  public Optional<JavaAnyOfComposition> getAnyOfComposition() {
    return anyOfComposition;
  }

  public PList<JavaRequiredAdditionalProperty> getRequiredAdditionalProperties() {
    return requiredAdditionalProperties;
  }

  public JavaAdditionalProperties getAdditionalProperties() {
    return additionalProperties;
  }

  public Constraints getConstraints() {
    return constraints;
  }

  public MemberGenerator.MemberContent getMemberContent() {
    return MemberContentBuilder.create()
        .isArrayPojo(false)
        .members(getAllMembers())
        .andAllOptionals()
        .additionalProperties(additionalProperties)
        .build();
  }

  public HashCodeGenerator.HashCodeContent getHashCodeContent() {
    return HashCodeContentBuilder.create()
        .members(getAllMembers())
        .andAllOptionals()
        .additionalProperties(additionalProperties)
        .build();
  }

  public EqualsGenerator.EqualsContent getEqualsContent() {
    return EqualsContentBuilder.create()
        .className(getClassName())
        .members(getAllMembers())
        .andAllOptionals()
        .additionalProperties(additionalProperties)
        .build();
  }

  public ToStringGenerator.ToStringContent getToStringContent() {
    return ToStringContentBuilder.create()
        .className(getClassName())
        .members(getAllMembers())
        .andAllOptionals()
        .additionalProperties(additionalProperties)
        .build();
  }

  public PojoConstructorGenerator.ConstructorContent getConstructorContent() {
    return ConstructorContentBuilder.create()
        .isArray(false)
        .className(getClassName())
        .members(getAllMembers())
        .andAllOptionals()
        .additionalProperties(additionalProperties)
        .build();
  }

  public WitherGenerator.WitherContent getWitherContent() {
    return WitherContentBuilder.create()
        .className(getClassName())
        .membersForWithers(getMembers())
        .membersForConstructorCall(getAllMembers())
        .andAllOptionals()
        .additionalProperties(additionalProperties)
        .build();
  }

  @Override
  public <T> T fold(
      Function<JavaArrayPojo, T> onArrayPojo,
      Function<JavaEnumPojo, T> onEnumPojo,
      Function<JavaObjectPojo, T> onObjectPojo) {
    return onObjectPojo.apply(this);
  }
}
