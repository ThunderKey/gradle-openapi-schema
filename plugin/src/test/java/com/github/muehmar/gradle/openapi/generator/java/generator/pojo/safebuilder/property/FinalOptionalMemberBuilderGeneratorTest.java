package com.github.muehmar.gradle.openapi.generator.java.generator.pojo.safebuilder.property;

import static com.github.muehmar.gradle.openapi.SnapshotUtil.writerSnapshot;
import static io.github.muehmar.codegenerator.writer.Writer.javaWriter;
import static org.junit.jupiter.api.Assertions.assertEquals;

import au.com.origin.snapshots.Expect;
import au.com.origin.snapshots.annotations.SnapshotName;
import au.com.origin.snapshots.junit5.SnapshotExtension;
import ch.bluecare.commons.data.PList;
import com.github.muehmar.gradle.openapi.generator.java.generator.pojo.safebuilder.SafeBuilderVariant;
import com.github.muehmar.gradle.openapi.generator.java.model.JavaAdditionalProperties;
import com.github.muehmar.gradle.openapi.generator.java.model.pojo.JavaObjectPojo;
import com.github.muehmar.gradle.openapi.generator.java.model.pojo.JavaPojos;
import com.github.muehmar.gradle.openapi.generator.settings.PojoSettings;
import com.github.muehmar.gradle.openapi.generator.settings.TestPojoSettings;
import io.github.muehmar.codegenerator.Generator;
import io.github.muehmar.codegenerator.writer.Writer;
import org.junit.jupiter.api.extension.ExtendWith;
import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.EnumSource;

@ExtendWith(SnapshotExtension.class)
class FinalOptionalMemberBuilderGeneratorTest {
  private Expect expect;

  @SnapshotName("allNecessityAndNullabilityVariants")
  @ParameterizedTest
  @EnumSource(SafeBuilderVariant.class)
  void generate_when_allNecessityAndNullabilityVariants_then_correctOutput(
      SafeBuilderVariant variant) {
    final Generator<JavaObjectPojo, PojoSettings> gen =
        FinalOptionalMemberBuilderGenerator.finalOptionalMemberBuilderGenerator(variant);

    final Writer writer =
        gen.generate(
            JavaPojos.allNecessityAndNullabilityVariants(),
            TestPojoSettings.defaultTestSettings(),
            javaWriter());

    expect.scenario(variant.name()).toMatchSnapshot(writerSnapshot(writer));
  }

  @ParameterizedTest
  @EnumSource(SafeBuilderVariant.class)
  @SnapshotName("noAdditionalPropertiesAllowed")
  void generate_when_noAdditionalPropertiesAllowed_then_noAdditionalPropertiesSetter(
      SafeBuilderVariant variant) {
    final Generator<JavaObjectPojo, PojoSettings> gen =
        FinalOptionalMemberBuilderGenerator.finalOptionalMemberBuilderGenerator(variant);

    final Writer writer =
        gen.generate(
            JavaPojos.objectPojo(PList.empty(), JavaAdditionalProperties.notAllowed()),
            TestPojoSettings.defaultTestSettings(),
            javaWriter());

    assertEquals(0, writer.getRefs().size());

    expect.scenario(variant.name()).toMatchSnapshot(writerSnapshot(writer));
  }
}
