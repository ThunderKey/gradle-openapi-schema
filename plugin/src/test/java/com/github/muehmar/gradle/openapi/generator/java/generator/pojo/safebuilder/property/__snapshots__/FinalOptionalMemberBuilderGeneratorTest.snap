allNecessityAndNullabilityVariants=[
java.util.Map

public static final class OptPropertyBuilder2 {
  private final Builder builder;

  private OptPropertyBuilder2(Builder builder) {
    this.builder = builder;
  }

  public OptPropertyBuilder2 addAdditionalProperty(String key, Object value) {
    return new OptPropertyBuilder2(builder.addAdditionalProperty(key, value));
  }

  public OptPropertyBuilder2 setAdditionalProperties(Map<String, Object> additionalProperties) {
    return new OptPropertyBuilder2(builder.setAdditionalProperties(additionalProperties));
  }

  public NecessityAndNullabilityDto build(){
    return builder.build();
  }
}
]


noAdditionalPropertiesAllowed=[
.
.
public static final class OptPropertyBuilder0 {
  private final Builder builder;

  private OptPropertyBuilder0(Builder builder) {
    this.builder = builder;
  }

  public ObjectPojo1Dto build(){
    return builder.build();
  }
}
]